src/Compilers/Shared/BuildServerConnection.cs

```diff
@@ -506,17 +507,28 @@ internal static string GetPipeNameForPathOpt(string compilerExeDirectory)
             return $"{userName}.{isAdmin}.{basePipeName}";
         }
 
-        internal static string GetBasePipeName(string compilerExeDirectory)
+        internal static string GetBasePipeName(string compilerExeDirectory, string sharedCompilationId = null)
         {
             // Normalize away trailing slashes.  File APIs include / exclude this with no 
             // discernable pattern.  Easiest to normalize it here vs. auditing every caller
             // of this method.
             compilerExeDirectory = compilerExeDirectory.TrimEnd(Path.DirectorySeparatorChar);
 
+            // Store refined compilerExeDirectory value as the basis for a raw pipe name to be hashed.
+            string rawPipeName = compilerExeDirectory;
+
+            // Append the sharedCompilationId value if it is provided.
+            if (!string.IsNullOrWhiteSpace(sharedCompilationId?.TrimStart('\0')))
+            {
+                // Trim away any whitespace from the sharedCompilationId.  Prepending a null terminator as "delimiter"
+                // between the path name and the sharedCompilationId.
+                rawPipeName += '\0' + sharedCompilationId.TrimStart('\0').Trim();
+            }
```

Comment:
This was a suggestion @phkelley made during an internal review; our thinking was to use a delimiter character to separate the path portion of the pipe name from the SharedCompilationId prior to hashing. This would protect against a few corner cases; specifically the following:

ExecutionDirectory: C:\fluf
SharedCompilationId: foo

Would have the same constructed pipe name as:

Execution Directory: C:\flu
SharedCompilationId: ffoo

We decided to go with a delimiter character that was unlikely/impossible to occur in either the Execution directory or SharedCompilationId; null terminator fit the bill. I doubt this case is likely to come up, but dealing with it is fairly low-cost.